File: dotnet/RailsLib.Net/Net/Game/State/PortfolioManager.cs
using GameLib.Net.Util;
using System;
using System.Collections.Generic;
using System.Text;

/**
 * PortfolioManager stores links to all existing portfolios
 */

namespace GameLib.Net.Game.State
{
    public class PortfolioManager : Manager
    {
        public class PMKey
        {
            private Type type;
            private IOwner owner;

            public PMKey(IPortfolio p)
            {
                this.type = p.PortfolioType;
                this.owner = p.PortfolioOwner;
            }

            public PMKey(Type t, IOwner owner)
            {
                this.type = t;
                this.owner = owner;
            }

            override public bool Equals(object other)
            {
                if (!(other is PMKey)) return false;
                PMKey otherKey = (PMKey)other;
                if (otherKey == null) return false;

                return type.Equals(otherKey.type) && owner.Equals(otherKey.owner);
            }

            override public int GetHashCode()
            {
                return type.GetHashCode() ^ owner.GetHashCode();
            }

            override public string ToString()
            {
                //return Objects.toStringHelper(this).add("Type", type).add("Owner", owner).toString();
                return $"{this.GetType().Name}{{Type={type}}}{{Owner={owner}}}";
            }

        }

        private DictionaryState<PMKey, IPortfolio> portfolios;

        private UnknownOwner unknown;


        private PortfolioManager(IItem parent, string id) : base(parent, id)
        {
            portfolios = DictionaryState<PMKey, IPortfolio>.Create(this, "portfolios");
            unknown = UnknownOwner.Create(this, "unknown");
        }

        public static PortfolioManager Create(StateManager parent, String id)
        {
            return new PortfolioManager(parent, id);
        }

        public UnknownOwner UnknownOwner
        {
            get
            {
                return unknown;
            }
        }

        /**
         * @param portfolio to add
         * @throws IllegalArgumentException if a portfolio of that type is already added
         */
        public void AddPortfolio<T>(Portfolio<T> portfolio) where T : IOwnable
        {
            PMKey key = new PMKey(portfolio);
            Precondition.CheckArgument(!portfolios.ContainsKey(key), "A portfolio of that type is defined for that owner already");
            portfolios.Put(key, portfolio);
        }

        /**
         * @param portfolio to remove
         */

        public void RemovePortfolio<T>(Portfolio<T> p) where T : IOwnable
        {
            portfolios.Remove(new PMKey(p));
        }

        /**
         * Returns the Portfolio that stores items of specified type for the specified owner
         * @param type class of items stored in portfolio
         * @param owner owner of the portfolio requested
         * @return portfolio for type/owner combination (null if none is available)
         */
        public Portfolio<T> GetPortfolio<T>(IOwner owner) where T : IOwnable
        {
            return (Portfolio<T>)portfolios.Get(new PMKey(typeof(T), owner));
        }

        // backdoor for testing
        public PMKey CreatePMKey<T>(IOwner owner) where T : IOwnable
        {
            return new PMKey(typeof(T), owner);
        }
    }
}

File: src/main/java/net/sf/rails/game/state/PortfolioManager.java
package net.sf.rails.game.state;

import com.google.common.base.MoreObjects;
import com.google.common.base.Objects;

import static com.google.common.base.Preconditions.checkArgument;

/**
 * PortfolioManager stores links to all existing portfolios
 */

public class PortfolioManager extends Manager {

    class PMKey<T extends Ownable> {
        private final Class<T> type;
        private final Owner owner;

        private PMKey(Portfolio<T> p) {
            this.type = p.getType();
            this.owner = p.getParent();
        }

        private PMKey(Class<T> type, Owner owner) {
            this.type = type;
            this.owner = owner;
        }

        @Override
        public boolean equals(Object other) {
            if (!(other instanceof PMKey)) return false;
            PMKey<?> otherKey = (PMKey<?>) other;
            return Objects.equal(type, otherKey.type) && Objects.equal(owner, otherKey.owner);
        }

        @Override
        public int hashCode() {
            return Objects.hashCode(type, owner);
        }

        @Override
        public String toString() {
            return MoreObjects.toStringHelper(this)
                    .add("Type", type)
                    .add("Owner", owner)
                    .toString();
        }

    }

    private final HashMapState<PMKey<? extends Ownable>, Portfolio<? extends Ownable>> portfolios =
            HashMapState.create(this, "portfolios");

    private final UnknownOwner unknown = UnknownOwner.create(this, "unknown");

    private PortfolioManager(Item parent, String id) {
        super(parent, id);
    }

    static PortfolioManager create(StateManager parent, String id) {
        return new PortfolioManager(parent, id);
    }

    UnknownOwner getUnkownOwner() {
        return unknown;
    }

    /**
     * @param portfolio to add
     * @throws IllegalArgumentException if a portfolio of that type is already added
     */
    <T extends Ownable> void addPortfolio(Portfolio<T> portfolio) {
        PMKey<T> key = new PMKey<T>(portfolio);
        checkArgument(!portfolios.containsKey(key), "A portfolio of that type is defined for that owner already");
        portfolios.put(key, portfolio);
    }

    /**
     * @param portfolio to remove
     */

    <T extends Ownable> void removePortfolio(Portfolio<T> p) {
        portfolios.remove(new PMKey<T>(p));
    }

    /**
     * Returns the Portfolio that stores items of specified type for the specified owner
     *
     * @param type  class of items stored in portfolio
     * @param owner owner of the portfolio requested
     * @return portfolio for type/owner combination (null if none is available)
     */
    // This suppress unchecked warnings is required as far I understand the literature on generics
    // however it should not be a problem as we store only type-safe portfolios
    @SuppressWarnings("unchecked")
    <T extends Ownable> Portfolio<T> getPortfolio(Class<T> type, Owner owner) {
        return (Portfolio<T>) portfolios.get(new PMKey<T>(type, owner));
    }

    // backdoor for testing
    <T extends Ownable> PMKey<T> createPMKey(Class<T> type, Owner owner) {
        return this.new PMKey<T>(type, owner);
    }

}
